torchok/tasks/detection.py

```diff
@@ -0,0 +1,106 @@
+        input_data = batch.get('image')
+        features = self.backbone.forward_features(input_data)[-self.num_scales:]
+        with torch.cuda.amp.autocast(dtype=torch.float32):
+            neck_out = self.neck(features)
+        prediction = self.bbox_head(neck_out)
+        output = self.bbox_head.format_dict(prediction)
+        output['image_shape'] = input_data.shape[-2:]
+
+        if 'bboxes' in batch:
+            output['gt_bboxes'] = batch.get('bboxes')
+            output['gt_labels'] = batch.get('label')
+
+        return output
+
+    def as_module(self) -> nn.Sequential:
+        """Method for model representation as sequential of modules(need for checkpointing)."""
+        return nn.Sequential(BackboneWrapper(self.backbone), self.neck, self.bbox_head)
+
+    def training_step(self, batch: Dict[str, Union[torch.Tensor, int]], batch_idx: int) -> Dict[str, torch.Tensor]:
+        """Complete training loop."""
+        output = self.forward_with_gt(batch)
+        total_loss, tagged_loss_values = self.bbox_head.loss(self.losses, **output)
+
+        output['prediction'] = self.bbox_head.get_bboxes(**output)
+        output['target'] = [dict(bboxes=bb, labels=la) for bb, la in zip(output['gt_bboxes'], output['gt_labels'])]
+        self.metrics_manager.update(Phase.TRAIN, **output)
+        output_dict = {'loss': total_loss}
+        output_dict.update(tagged_loss_values)
+        return output_dict
+
+    def validation_step(self, batch: Dict[str, Union[torch.Tensor, int]],
+                        batch_idx: int, dataloader_idx: int = 0) -> Dict[str, torch.Tensor]:
+        """Complete validation loop."""
+        output = self.forward_with_gt(batch)
+        total_loss, tagged_loss_values = self.bbox_head.loss(self.losses, **output)
+
+        output['prediction'] = self.bbox_head.get_bboxes(**output)
+        output['target'] = [dict(bboxes=bb, labels=la) for bb, la in zip(output['gt_bboxes'], output['gt_labels'])]
+        self.metrics_manager.update(Phase.VALID, **output)
+        output_dict = {'loss': total_loss}
+        output_dict.update(tagged_loss_values)
+        return output_dict
+
+    def test_step(self, batch: Dict[str, Union[torch.Tensor, int]], batch_idx: int) -> None:
+        """Complete test loop."""
+        output = self.forward_with_gt(batch)
+        output['prediction'] = self.bbox_head.get_bboxes(**output)
+        output['target'] = [dict(bboxes=bb, labels=la) for bb, la in zip(output['gt_bboxes'], output['gt_labels'])]
+        self.metrics_manager.update(Phase.TEST, **output)
+
+    def predict_step(self, batch: Dict[str, Union[torch.Tensor, int]], batch_idx: int) -> Dict[str, torch.Tensor]:
+        """Complete predict loop."""
+        output = self.forward_with_gt(batch)
+        output['prediction'] = self.bbox_head.get_bboxes(**output)
+        output['target'] = [dict(bboxes=bb, labels=la) for bb, la in zip(output['gt_bboxes'], output['gt_labels'])]
```

Comment:
Need to check if output has gt_bboxes and gt_labels or remove this line, because it's predict

Reply:
Fixed